File: Ex4.java
package Ex4;

abstract class Animal {
    public abstract void som();
}

class Cachorro extends Animal {
    @Override
    public void som() {
        System.out.println("O cachorro faz: Au Au!");
    }
}

class Gato extends Animal {
    @Override
    public void som() {
        System.out.println("O gato faz: Miau!");
    }
}

public class Ex4 {
    public static void main(String[] args) {
        Animal meuCachorro = new Cachorro();
        Animal meuGato = new Gato();
        
        meuCachorro.som();
        meuGato.som();
    }
}

File: ex4.go
package main

import (
	"fmt"
)

type Animal interface {
	Som() string
}

type Cachorro struct{}

func (c Cachorro) Som() string {
	return "Au Au"
}

type Gato struct{}

func (g Gato) Som() string {
	return "Miau"
}

func ExibirSom(animal Animal) {
	fmt.Printf("O animal faz: %s\n", animal.Som())
}

func main() {
	var animal1 Animal = Cachorro{}
	var animal2 Animal = Gato{}

	ExibirSom(animal1)
	ExibirSom(animal2)
}
